### PR TITLE
[ESP32] Read core dump from flash, for diagnostic crash logs

### LINKED ISSUE #31727: [ESP32] Report the actual crash logs or core dump from flash for diagnostic logs
https://github.com/project-chip/connectedhomeip/pull/31682 adds the implementation of diagnostic log provider delegate. It reads the data from the files embedded in firmware.

ESP32 can store the crash on the flash and we can retrieve the crash on the next boot.

This is to track sending an actual crash instead of some dummy logs.

### PR DESCRIPTION
**Pull request by shubhamdp**
Send actual core dump as diagnostic crash logs.

Fixes #31727

#### Tests
Crashed device once and read and decoded the core dump

### CODE REVIEW

File: examples/temperature-measurement-app/esp32/main/include/diagnostic-logs-provider-delegate-impl.h [hunk 1]
```diff
@@ -54,12 +55,59 @@ class LogProvider : public DiagnosticLogsProviderDelegate
     LogProvider(const LogProvider &)             = delete;
     LogProvider & operator=(const LogProvider &) = delete;
 
-    // This tracks the ByteSpan for each session
-    std::map<LogSessionHandle, ByteSpan *> mSessionSpanMap;
+    struct CrashLogContext
+    {
+        spi_flash_mmap_handle_t mappedHandle;
+        const void * mappedAddress;
```
Comment (yunhanw-google) [inline examples/temperature-measurement-app/esp32/main/include/diagnostic-logs-provider-delegate-impl.h:61]: please init these values in struct

File: examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp [hunk 2]
```diff
@@ -69,93 +75,232 @@ CHIP_ERROR LogProvider::GetLogForIntent(IntentEnum intent, MutableByteSpan & out
...
-    auto dataSize   = span->size();
-    auto count      = std::min(dataSize, outBuffer.size());
-
-    VerifyOrReturnError(CanCastTo<off_t>(count), CHIP_ERROR_INVALID_ARGUMENT, outBuffer.reduce_size(0));
-
-    ReturnErrorOnFailure(CopySpanToMutableSpan(ByteSpan(span->data(), count), outBuffer));
+    VerifyOrReturnValue(mSessionContextMap.count(sessionHandle), CHIP_ERROR_INVALID_ARGUMENT);
 
-    outIsEndOfLog = dataSize == count;
+    LogContext * context = mSessionContextMap[sessionHandle];
```
Comment (yunhanw-google) [inline examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp:302]: please check null?

File: examples/temperature-measurement-app/esp32/main/include/diagnostic-logs-provider-delegate-impl.h (hunk 1 above)
Comment (shubhamdp) [reply examples/temperature-measurement-app/esp32/main/include/diagnostic-logs-provider-delegate-impl.h]: I had them, but after adding Reset() I though that would be redundant. Adding them back.

File: examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp (hunk 2 above)
Comment (shubhamdp) [reply examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp:303]: Thanks for bringing it up, added the null checks at necessary places.

**Review by dhrishi — APPROVED** [review]

File: examples/temperature-measurement-app/esp32/main/include/diagnostic-logs-provider-delegate-impl.h (hunk 1 above)
Comment (yunhanw-google) [reply examples/temperature-measurement-app/esp32/main/include/diagnostic-logs-provider-delegate-impl.h]: yep, usually we prefer init them in members directly. thanks

**Review by yunhanw-google — APPROVED** [review]

**Comment by dhrishi** [comment]
@woody-apple @andy31415 Can you please cherry-pick this platform specific change to `1.3-sve` branch?

File: examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp [hunk 3]
```diff
@@ -69,93 +75,233 @@ CHIP_ERROR LogProvider::GetLogForIntent(IntentEnum intent, MutableByteSpan & out
...
+            // reduce the span after reading count bytes
+            context->NetworkDiag.span = context->NetworkDiag.span.SubSpan(count);
+        }
+    }
+    break;
+
+    case IntentEnum::kCrashLogs: {
+#if defined(CONFIG_ESP_COREDUMP_ENABLE_TO_FLASH) && defined(CONFIG_ESP_COREDUMP_DATA_FORMAT_ELF)
+        CrashLogContext * logContext = context->Crash.logContext;
+        size_t dataSize              = logContext->crashSize - logContext->readOffset;
```
[the remark this answers is not in the document]
Comment (dhrishi) [reply examples/temperature-measurement-app/esp32/main/diagnostic-logs-provider-delegate-impl.cpp:230]: @shubhamdp

**Pull request merged**